### PR TITLE
Retry fetching and publishing records

### PR DESCRIPTION
**Pull request by leplatrem**

### CODE REVIEW

File: jobs/scrape_archives.py [hunk 1]
```diff
@@ -16,20 +17,21 @@
 DEFAULT_SERVER = "https://kinto-ota.dev.mozaws.net/v1"
 DEFAULT_BUCKET = "build-hub"
 DEFAULT_COLLECTION = "archives"
-NB_THREADS = 5
+NB_THREADS = 3
+TIMEOUT = 30
```
Comment (Natim) [inline jobs/scrape_archives.py]: TIMEOUT_SECONDS

File: jobs/scrape_archives.py [hunk 2]
```diff
@@ -43,12 +44,17 @@ def archive_url(product, version=None, platform=None, locale=None):
...
+                       asyncio.TimeoutError,
+                       concurrent.futures.CancelledError),
+                      max_tries=8)
 async def fetch_listing(session, url):
     headers = {
         "Accept": "application/json",
         "User-Agent": "BuildHub;storage-team@mozilla.com"
     }
-    async with session.get(url, headers=headers) as response:
+    async with session.get(url, headers=headers, timeout=30) as response:
```
Comment (Natim) [inline jobs/scrape_archives.py]: Use the constants

File: jobs/scrape_archives.py [hunk 3]
```diff
@@ -159,6 +168,7 @@ def main():
         description="Send releases archives to Kinto",
         default_server=DEFAULT_SERVER,
         default_bucket=DEFAULT_BUCKET,
+        default_retry=3,
```
Comment (Natim) [inline jobs/scrape_archives.py]: Use a constant maybe?

File: jobs/scrape_archives.py [hunk 4]
```diff
@@ -43,12 +45,17 @@ def archive_url(product, version=None, platform=None, locale=None):
     return url
 
 
+@backoff.on_exception(backoff.expo,
+                      (aiohttp.ClientError,
+                       asyncio.TimeoutError,
+                       concurrent.futures.CancelledError),
+                      max_tries=8)
```
Comment (Natim) [inline jobs/scrape_archives.py]: Why 8 there and 3 below?

**Comment by leplatrem** [comment]
It is not robust at all, there are many errors while running the script:

```
Exception ignored in: <coroutine object fetch_files at 0x7fb7bc941ca8>
Traceback (most recent call last):
  File "scrape_archives.py", line 110, in fetch_files
    _, files = await fetch_listing(session, locale_url)
  File "/home/mathieu/Code/Mozilla/buildhub/jobs/.venv3/lib/python3.5/site-packages/backoff/_async.py", line 120, in retry
    ret = yield from target(*args, **kwargs)
  File "scrape_archives.py", line 58, in fetch_listing
    async with session.get(url, headers=headers, timeout=TIMEOUT) as response:
  File "/home/mathieu/Code/Mozilla/buildhub/jobs/.venv3/lib/python3.5/site-packages/aiohttp/client.py", line 626, in __aenter__
    self._resp = yield from self._coro
  File "/home/mathieu/Code/Mozilla/buildhub/jobs/.venv3/lib/python3.5/site-packages/aiohttp/client.py", line 225, in _request
    conn = yield from self._connector.connect(req)
  File "/home/mathieu/Code/Mozilla/buildhub/jobs/.venv3/lib/python3.5/site-packages/aiohttp/connector.py", line 382, in connect
    self._acquired.remove(placeholder)
KeyError: <aiohttp.connector._TransportPlaceholder object at 0x7fb788ef3898>
_GatheringFuture exception was never retrieved
```
```
future: <_GatheringFuture finished exception=RuntimeError('Session is closed',)>
Traceback (most recent call last):
  File "/home/mathieu/Code/Mozilla/buildhub/jobs/.venv3/lib/python3.5/site-packages/aiohttp/client.py", line 225, in _request
    conn = yield from self._connector.connect(req)
  File "/home/mathieu/Code/Mozilla/buildhub/jobs/.venv3/lib/python3.5/site-packages/aiohttp/connector.py", line 364, in connect
    yield from fut
  File "/usr/lib/python3.5/asyncio/futures.py", line 380, in __iter__
    yield self  # This tells Task to wait for completion.
  File "/usr/lib/python3.5/asyncio/tasks.py", line 304, in _wakeup
    future.result()
  File "/usr/lib/python3.5/asyncio/futures.py", line 285, in result
    raise CancelledError
concurrent.futures._base.CancelledError
```
```During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.5/asyncio/tasks.py", line 239, in _step
    result = coro.send(None)
  File "scrape_archives.py", line 110, in fetch_files
    _, files = await fetch_listing(session, locale_url)
  File "/home/mathieu/Code/Mozilla/buildhub/jobs/.venv3/lib/python3.5/site-packages/backoff/_async.py", line 120, in retry
    ret = yield from target(*args, **kwargs)
  File "scrape_archives.py", line 58, in fetch_listing
    async with session.get(url, headers=headers, timeout=TIMEOUT) as response:
  File "/home/mathieu/Code/Mozilla/buildhub/jobs/.venv3/lib/python3.5/site-packages/aiohttp/client.py", line 626, in __aenter__
    self._resp = yield from self._coro
  File "/home/mathieu/Code/Mozilla/buildhub/jobs/.venv3/lib/python3.5/site-packages/aiohttp/client.py", line 164, in _request
    raise RuntimeError('Session is closed')
RuntimeError: Session is closed
```

File: jobs/scrape_archives.py (hunk 4 above)
Comment (leplatrem) [inline jobs/scrape_archives.py]: you're right we could get rid of the backoff for kinto operations since we already have retry in client

**Comment by Natim** [comment]
I guess we should be using a session per request.

File: jobs/scrape_archives.py [hunk 5]
```diff
@@ -146,9 +153,11 @@ def archive_url(product, version=None, platform=None, locale=None):
...
     batch_size = info["settings"]["batch_max_requests"]
     while True:
+        # Gather a batch of records.
         records = [await queue.get()]
         while not queue.empty() and len(records) < batch_size:
             records.append(await queue.get())
+
         task = loop.run_in_executor(executor, publish_records, client, records)
         # Counting tasks done allows us to wait for the queue with queue.join()
         task.add_done_callback(lambda fut: queue.task_done())
```
Comment (leplatrem) [inline jobs/scrape_archives.py]: Maybe here we should call `task_done()`  as many time as created records in batch?

**Comment by leplatrem** [comment]
I still have 
```
Future exception was never retrieved
future: <Future finished exception=KintoException(KintoException(...), "409 - {'message': 'Integrity constraint violated, please retry.', 'errno': 122, 'error': 'Conflict', 'code': 409}")>
Traceback (most recent call last):
  File "/usr/lib/python3.5/concurrent/futures/thread.py", line 55, in run
    result = self.fn(*self.args, **self.kwargs)
  File "scrape_archives.py", line 31, in publish_records
    batch.create_record(record)
  File "/usr/lib/python3.5/contextlib.py", line 66, in __exit__
    next(self.gen)
  File "/home/mathieu/Code/Mozilla/kinto.py/kinto_http/__init__.py", line 101, in batch
    batch_session.send()
  File "/home/mathieu/Code/Mozilla/kinto.py/kinto_http/batch.py", line 60, in send
    raise exception
kinto_http.exceptions.KintoException: (KintoException(...), "409 - {'message': 'Integrity constraint violated, please retry.', 'errno': 122, 'error': 'Conflict', 'code': 409}")
```

**Comment by leplatrem** [comment]
This seems to be the right approach.

**Comment by leplatrem** [comment]
This version works well, merging.

**Pull request merged**